### PR TITLE
Update DS3231 binding to add support for the alarms

### LINKED ISSUE #1031: Add support for alarms on the DS3231 RTC
The DS3231 RTC implementation (in devices/Rtc) at present only includes functionality to get the time, set the time and get the temperature. The device itself also has the ability to set two separate alarms that will fire when the specified time is matched (according to various matching modes), which I am in need of for a project I am working on.

[This](https://github.com/Makuna/Rtc/) repo, which I have used in other projects, provides a C implementation of the DS3231 (including alarms). The relevant parts for alarms in that repo (see src/RtcDS3231.h) are anything to do with alarm one/two, square wave, and latch. Note: I'm not sure how the license of that repo affects the ability to utilise it in whatever way here, but I thought I'd reference it just to show what needs implementing.

Personally I only need alarm one working with the once-per-second mode.

### PR DESCRIPTION
**Pull request by henryshunt**
Fixes #1031 

Adds support for both of the built-in alarms on the DS3231 RTC chip.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1150) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1150) before we can accept your contribution.<br/><br/>:x: henryshunt [sign now](https://cla.dotnetfoundation.org/dotnet/iot?pullRequest=1150)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/iot?pullRequest=1150) it.</sub>

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs
Comment (henryshunt) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs:10]: I don't recall the datasheet saying that both alarms cannot be active at the same time, and there are sources saying that they can both be active, however I was unable to get both to work at the same time. Both alarms worked fine independently, but as soon as they were both enabled, neither alarm would trigger. Took quite some time to figure that out.

I can have another look if you'd like, but I remember looking at the raw values of all registers to confirm that everything was correct, and that nothing except the two alarm enabled bits were changing between enabling one alarm and enabling both.

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs
Comment (henryshunt) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs:10]: The settings aren't the same. Alarm 1 is comprised of a day, hour, minute, second and match mode. Alarm 2 is comprised of a day, hour, minute and match mode. The match modes for each alarm are also different.

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs
Comment (pgrawehr) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs:10]: That's odd then. Maybe make it a flag anyway and note the above in the comment? The API should reflect the ideal case of how things should work, so that any underlying bugs (be it our implementation or the physical layer itself) can later be fixed without an api change.

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs
Comment (pgrawehr) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs:10]: I think the user should most of the time not have to worry about that. Throw an exception if one attempts to do something unsupported (ie. use alarm 2 with seconds), but otherwise keep these differences internal. You could change these classes to internal and use a public, common one in the public API.

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs
Comment (henryshunt) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs:10]: Apologies for the delayed response. I've been too busy to spend any time on this project but I will definitely get back to you on these two issues when I have the time again.

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs
Comment (krwq) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs:10]: (internally you can have two classes but as public API I agree one would be much easier to use)

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs
Comment (krwq) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs:10]: I think this should be internal as well

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs
Comment (krwq) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs:10]: If you feel they're distinct enough though please make sure to rename them accordingly, 1,2 suffix is not distinct enough and suggests similarity

File: src/devices/Rtc/Devices/Ds3231/Ds3231.cs
Comment (krwq) [inline src/devices/Rtc/Devices/Ds3231/Ds3231.cs:194]: Consider: instead of Set/Read alarm methods have a nullable property and immutable class with a single instance of this class i.e.:

```csharp
Ds3231 x = ...;
x.Alarm = Ds3231.CreateAlarm(...); // you can have two similar methods for CreateAlarm to cover both cases (perhaps different name to better describe it, constructor is fine as well if args are different enough)
```

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs
Comment (krwq) [inline src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs:3]: use 2 line copyright header

**Review by krwq — APPROVED** [review]

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm1.cs
Comment (Ellerbach) [inline src/devices/Rtc/Devices/Ds3231/Ds3231Alarm1.cs:15]: Why not use a TimeStamp for the Hour, Days, etc? 
That would save you a lot of checks.

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs
Comment (Ellerbach) [inline src/devices/Rtc/Devices/Ds3231/Ds3231Alarm2.cs:20]: Same here. TimeStamp looks good for Hours, Minutes, Seconds

**Review by Ellerbach — COMMENTED** [review]
For the Hours, Minutes, Seconds, a TimeStamp seems better to me than the individual elements.

File: src/devices/Rtc/Devices/Ds3231/Ds3231.cs
Comment (krwq) [inline src/devices/Rtc/Devices/Ds3231/Ds3231.cs:126]: did you mean time stamp or time span?

Comment (krwq) [reply src/devices/Rtc/Devices/Ds3231/Ds3231.cs:126]: seems we don't have any existing suitable structure. fine as is with me

**Review by krwq — APPROVED** [review]

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs
Comment (Ellerbach) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs:10]: It seems that both alarms can be setup together. But I can't test as I don't have the device. So leaving it like it is.

File: src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs
Comment (Ellerbach) [reply src/devices/Rtc/Devices/Ds3231/Ds3231Alarm.cs:3]: Adjusting in this commit as well

**Comment by joperezr** [comment]
[Triage]: This is good to go after fixing merge conflict.

**Pull request merged**
